core/schema/modulesource.go

```diff
@@ -0,0 +1,1056 @@
+		}
+		for _, path := range localDep.modCfg.Exclude {
+			absPath := filepath.Join(sourceRootAbsPath, path)
+			relPath, err := filepath.Rel(contextAbsPath, absPath)
+			if err != nil {
+				return inst, fmt.Errorf("failed to get relative path of config exclude: %s", err)
+			}
+			if !filepath.IsLocal(relPath) {
+				return inst, fmt.Errorf("local module dep source exclude path %q escapes context %q", relPath, contextAbsPath)
+			}
+			excludeSet[relPath] = struct{}{}
+		}
+
+		// always include the config file
+		configRelPath, err := filepath.Rel(contextAbsPath, filepath.Join(rootPath, modules.Filename))
+		if err != nil {
+			return inst, fmt.Errorf("failed to get relative path: %s", err)
+		}
+		includeSet[configRelPath] = struct{}{}
+
+		// always include the source dir
+		source := localDep.modCfg.Source
+		if source == "" {
+			source = "."
+		}
+		sourceAbsSubpath := filepath.Join(rootPath, source)
+		sourceRelSubpath, err := filepath.Rel(contextAbsPath, sourceAbsSubpath)
+		if err != nil {
+			return inst, fmt.Errorf("failed to get relative path: %s", err)
+		}
+		if !filepath.IsLocal(sourceRelSubpath) {
+			return inst, fmt.Errorf("local module source path %q escapes context %q", sourceRelSubpath, contextAbsPath)
+		}
+		includeSet[sourceRelSubpath+"/**/*"] = struct{}{}
+	}
+
+	for _, sdk := range sdkSet {
+		// NOTE: required paths are currently **-style globs that apply to the whole context subtree
+		// This is a bit of a delicate assumption, if need arises for including exact paths, this
+		// will need some adjustment.
+		requiredPaths, err := sdk.RequiredPaths(ctx)
+		if err != nil {
+			return inst, fmt.Errorf("failed to get sdk required paths: %w", err)
+		}
+		for _, path := range requiredPaths {
+			includeSet[path] = struct{}{}
+		}
+	}
+
+	includes := make([]string, 0, len(includeSet))
+	for include := range includeSet {
+		includes = append(includes, include)
+	}
+	excludes := make([]string, 0, len(excludeSet))
+	for exclude := range excludeSet {
+		excludes = append(excludes, exclude)
+	}
+
+	pipelineName := fmt.Sprintf("load local module context %s", contextAbsPath)
+	ctx, subRecorder := progrock.WithGroup(ctx, pipelineName, progrock.Weak())
+	_, desc, err := src.Query.Buildkit.LocalImport(
+		ctx, subRecorder, src.Query.Platform.Spec(),
+		contextAbsPath,
+		excludes,
+		includes,
+	)
+	if err != nil {
+		return inst, fmt.Errorf("failed to import local module source: %w", err)
+	}
+	loadedDir, err := core.LoadBlob(ctx, s.dag, desc)
+	if err != nil {
+		return inst, fmt.Errorf("failed to load local module source: %w", err)
+	}
+
+	return s.normalizeCallerLoadedSource(ctx, src, sourceRootRelPath, loadedDir)
+}
+
+// get an instance of ModuleSource with the context resolved from the caller that doesn't
+// encode any instructions to actually reload from the caller if the ID is loaded later, which
+// is possible due to blob-ifying the local import.
+func (s *moduleSchema) normalizeCallerLoadedSource(
```

Comment:
Ah so this is the tricky part. Definitely a bit awkward, but this is the price you pay for reproducible pipelines, I guess.

Does this only affect local module sources? For example, if I ran with `-m` pointing to a Git ref, will the resulting ID's module still reference `git` like normal and keep all the nice theoretical portability properties? (I think the answer is yes, but just making sure.)

Reply:
> Does this only affect local module sources? For example, if I ran with -m pointing to a Git ref, will the resulting ID's module still reference git like normal and keep all the nice theoretical portability properties? (I think the answer is yes, but just making sure.)

Yep only local sources right now. Git sources are inherently normalized like you said, much easier to deal with 😄